aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -326,6 +360,123 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    public entry fun rotate_authentication_key_with_rotation_capability(
+        delegate_signer: &signer,
+        rotation_cap_offerer_address: address,
+        new_scheme: u8,
+        new_public_key_bytes: vector<u8>,
+        cap_update_table: vector<u8>
+    ) acquires Account, OriginatingAddress {
+        assert!(exists_at(rotation_cap_offerer_address), error::not_found(EOFFERER_ADDRESS_DOES_NOT_EXIST));
+
+        // check that there exists a rotation capability offer at the offerer's account resource for the delegate
+        let offerer_account_resource = borrow_global<Account>(rotation_cap_offerer_address);
+        let delegate_address = signer::address_of(delegate_signer);
+        assert!(option::contains(&offerer_account_resource.rotation_capability_offer.for, &delegate_address), error::not_found(ENO_SUCH_ROTATION_CAPABILITY_OFFER));
+
+        let delegate_account_resource = borrow_global<Account>(delegate_address);
+        let curr_auth_key = from_bcs::to_address(offerer_account_resource.authentication_key);
+        let challenge = RotationProofChallenge {
+            sequence_number: delegate_account_resource.sequence_number,
+            originator: rotation_cap_offerer_address,
+            current_auth_key: curr_auth_key,
+            new_public_key: new_public_key_bytes,
+        };
+
+        // verify that `cap_update_table` has the correct `RotationProofChallenge` and is signed by the new private key
+        let new_auth_key = assert_valid_signature_and_get_auth_key(new_scheme, new_public_key_bytes, cap_update_table, &challenge);
+
+        // update the `OriginatingAddress` table, so we can find the originating address using the new address
+        let new_address = from_bcs::to_address(new_auth_key);
```

Comment:
So, the code below that updates the `OriginatingAddress` table is now a duplicate of the code in `rotate_authentication_key`.

This makes me uneasy because future changes might miss updating the code in both places.

Can we please refactor and introduce a function? Something like:

```
/// Checks `OriginatingAddress[curr_auth_key]` is either unmapped, or mapped to `originating_address`. If it's mapped, it removes that mapping. Finally, always sets `OriginatingAddress[new_auth_key] = originating_address`.
fun update_originating_address_table(
   originating_addr: address,
   curr_auth_key: address,
   new_auth_key: address,
)
```
You would then delete this code and instead call `update_originating_address_table(rotation_cap_offerer_address, curr_auth_key, new_address)` .

You would also have to update `rotate_authentication_key` to call `update_originating_address_table(addr, curr_auth_key_as_address, new_auth_key_as_address)`.

Reply:
While, we're at it, the `OriginatingAddress` struct has no comments explaining what it does.

Can we modify it to some variant of the following:
```
/// It is easy to fetch the authentication key of an address by simply reading it from the `Account` struct at that address.
/// The table in this struct makes it possible to do a reverse lookup: it maps an authentication key, to the address of the account which has that authentication key set.
///
/// This mapping is needed when recovering wallets for accounts whose authentication key has been rotated.
///
/// For example, imagine a freshly-created wallet with address `a` and thus also with authentication key `a`, derived from a PK `pk_a` with corresponding SK `sk_a`. 
/// It is easy to recover such a wallet given just the secret key `sk_a`, since the PK can be derived from the SK, the authentication key can then be derived from the PK, and the address equals the authentication key (since there was no key rotation).
///
/// However, if such a wallet rotates its authentication key to `b` derived from a different PK `pk_b` with SK `sk_b`, how would account recovery work?
/// The recovered address would no longer be 'a'; it would be `b`, which is incorrect.
/// This struct solves this problem by mapping the new authentication key `b` to the original address `a` and thus helping the wallet software during recovery find the correct address.
struct OriginatingAddress has key {
    address_map: Table<address, address>,
}
```

Reply:
updated!

Reply:
Cool, thank you!

One potentially-pedantic point (I don't know): You are reading the account resource a second time inside `update_originating_address_table` since you did not chose to give the current authentication key as a parameter.

Do we want to do this? Should we not make the code gas-efficient, if it's easy to do so?

Reply:
ahh so I thought we also want to update the account's auth key in this function, if not we can just pass in the auth key then. will update soon